lib/charms/observability_libs/v0/cert_manager.py

```diff
@@ -0,0 +1,322 @@
+
+
+class CertManagerEvents(ObjectEvents):
+    cert_changed = EventSource(CertChanged)
+
+
+class CertManager(Object):
+    """CertManager is used to wrap TLS Certificates management operations for charms.
+
+    TODO: figure out if the constructor should take
+        key_path: str,
+        cert_path: str,
+        ca_path: str,
+     instead of charm code pushing & deleting them itself.
+
+    TODO Use some sort of compound status
+
+    CertManager manages one single cert.
+    """
+    on = CertManagerEvents()  # pyright: ignore
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        *,
+        peer_relation_name: str,
+        certificates_relation_name: str = "certificates",
+        cert_subject: Optional[str] = None,
+        key: str = "cert-manager",  # TODO what to put here?
+    ):
+        super().__init__(charm, key)
+
+        self.charm = charm
+        app_name = charm.unit.name.split("/")[0]
+        self.cert_subject = app_name if not cert_subject else cert_subject
+        self.peer_relation_name = peer_relation_name
+        self.certificates_relation_name = certificates_relation_name
+
+        self.certificates = TLSCertificatesRequiresV2(self.charm, self.certificates_relation_name)
+
+        self.framework.observe(
+            self.charm.on.certificates_relation_joined,  # pyright: ignore
+            self._on_certificates_relation_joined,
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_available,  # pyright: ignore
+            self._on_certificate_available,
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_expiring,  # pyright: ignore
+            self._on_certificate_expiring,
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_invalidated,  # pyright: ignore
+            self._on_certificate_invalidated,
+        )
+        self.framework.observe(
+            self.certificates.on.all_certificates_invalidated,  # pyright: ignore
+            self._on_all_certificates_invalidated,
+        )
+
+        # Peer relation events
+        self.framework.observe(
+            self.charm.on[self.peer_relation_name].relation_created, self._on_peer_relation_created
+        )
+
+    @property
+    def _peer_relation(self) -> Optional[Relation]:
+        """Return the peer relation."""
+        return self.charm.model.get_relation(self.peer_relation_name, None)
+
+    def _on_peer_relation_created(self, _):
+        """Generate the private key and store it in a peer relation."""
+        if not self.charm.unit.is_leader():
+            return
+
+        # We're in "relation-joined", so the relation should be there
+        peer_relation = self._peer_relation
+
+        # Just in case we already have a private key, do not overwrite it.
+        # Not sure how this could happen.
+        # TODO figure out how to go about key rotation.
```

Comment:
I would maybe add some utility function that does this; wherever the input for key rotation comes from (a charm action?):

```python
def rotate_keys(self, _) -> None:
    """Generate a new private key and certificate for key rotation, revoking the old one."""
    if not self.charm.unit.is_leader() or not self._peer_relation:
        return
        
    private_key = generate_private_key(self.private_key_password)
    self._private_key = private_key.decode()
    
    self._certificate_revoked()
    self._generate_csr()
```

Generate a new key, revoke the current certificate and ask for a new one seems reasonable